### PR TITLE
[For discussion] Camera access

### PR DESCRIPTION
**Pull request by huningxin**

### CODE REVIEW

**Comment by blairmacintyre** [comment]
@huningxin sorry to drop this.  Was busy with other things and missed the notification go by.  @TrevorFSmith and I will look at it ASAP!

**Comment by blairmacintyre** [comment]
So, my take on it @TrevorFSmith was that the act of asking for devices would trigger the UA to question the user as to if the app can have access to the cameras. I would want to discuss with the security folks what info is available and what isn't available with user permission.  BUT, I wouldn't expect to be able to access the actual camera data without user permission -- no user-defined CV without the UA giving approval (just like access to the camera now).

**Comment by blairmacintyre** [comment]
Question:  do you think we should worry about cases where the platform can provide video in a WebGL texture, but not in main memory?  Should we make the imagebitmap optionally null, and add a texture?

**Comment by blairmacintyre** [comment]
Another question:  does providing the data inside the frame means the CV happens inside the rAF callback?  If we set up separate callbacks, or workers, could we get some parallelization or get the CV going before the rAF is ready to fire?  For example, if the frame rate of the video cameras are different than the rendering frame rate, shouldn't we be running CV on the camera ASAP, decoupled from the rAF?  How would we do that?

I'm thinking about future/modern browsers with workers that support parallelization.

File: WebXR API.md [hunk 1]
```diff
@@ -79,6 +81,10 @@ A Hololens could expose a single passthrough display.
 		long requestFrame(XRFrameRequestCallback callback);
 		void cancelFrame(long handle);
 
+		// Request the camera frame available in XRPresentationFrame.
+		// This operation will ask for users' permission.
+		Promise<void> requestCameraAccess(XRCameraSource camera);
```
Comment (huningxin) [inline WebXR API.md]: Request camera frame will ask for users' permission. Without permission, the camera frame will not be available in presentation frame.

File: WebXR API.md [hunk 2]
```diff
@@ -198,6 +204,42 @@ XRAnchorOffset represents a position in relation to an anchor, returned when XRP
...
+
+## XRCameraFrame
+
+	interface XRCameraFrame {
+		readonly attribute ImageBitmap capturedImage;
+		readonly attribute Float32Array projectionMatrix;
+		XRCameraPose? getPose(XRCoordinateSystem coordinateSystem); 
+	};
+	
+[ImageBitmap Extension](https://w3c.github.io/mediacapture-worker/#imagebitmap-extensions) allows developers to process the captured image in native format, e.g. YUV, RGB or DEPTH, by either asm.js/wasm or WebGL/WebGPU.
```
Comment (huningxin) [inline WebXR API.md]: > Question: do you think we should worry about cases where the platform can provide video in a WebGL texture, but not in main memory?

@blairmacintyre , the ImageBitmap Extension supports GPU processing as well.

File: WebXR API.md [hunk 3]
```diff
@@ -228,6 +272,10 @@ _The XRPresentationFrame differs from the VRPresentationFrame with the addition
 - Should we fire an event when a marker or feature based anchor (e.g. a wall, a table top) is detected?
 - How can we access camera image buffers aor textures?
 
+`getCameraFrame` allows developers to access camera image and pose data. For some devices, the camera frame is not always available in each presentation frame. `getCameraFrame` returns `XRCameraFrame` object when sychronized camera frame is available to current presentation frame, otherwise, it returns nothing. 
+
+When `XRCameraFrame` available, developers are able to draw the camera image by WebGL insdie the rendering loop for devices. Or developers are able to send the camera frame to a worker thread for computer vision processing, e.g. marker detection, in parallel.
+
```
Comment (huningxin) [inline WebXR API.md]: > Another question: does providing the data inside the frame means the CV happens inside the rAF callback? 

@blairmacintyre , the CV happens on worker thread. If developers want to render the camera image, it happens inside the rAF callback.

**Comment by huningxin** [comment]
@blairmacintyre @TrevorFSmith , thanks for the inputs!

> the act of asking for devices would trigger the UA to question the user as to if the app can have access to the cameras

Exactly, `vrSession.requestCameraFrame` will ask for users' permission.

> Question: do you think we should worry about cases where the platform can provide video in a WebGL texture, but not in main memory?

The [ImageBitmap Extension](https://w3c.github.io/mediacapture-worker/#imagebitmap-extensions) supports efficient image accessing for both CPU (asm.js/wasm) and GPU (WebGL):

quotation:
1. If developers use JavaScript(/asm.js) to process the frames, then the ImageBitmap interface needs to be extended with APIs for developers to access its underlying data and there should also be a way for developers to create an ImageBitmap from the processed data.
2. If developers use WebGL, then WebGL needs to be extended so that developers can pass an ImageBitmap into the WebGL context and the browser will handle how to upload the raw image data into the GPU memory. Possibly, the data is already in the GPU memory so that the operation could be very efficient.

> Another question: does providing the data inside the frame means the CV happens inside the rAF callback? If we set up separate callbacks, or workers, could we get some parallelization or get the CV going before the rAF is ready to fire? For example, if the frame rate of the video cameras are different than the rendering frame rate, shouldn't we be running CV on the camera ASAP, decoupled from the rAF? How would we do that?

I used to think sync the camera frame to the presentation frame by get the camera frame inside presentation frame. However it doesn't allow us to process the camera frame ahead of presentation frame. I added `XRCameraFrameProcessor` interface to decouple the camera frame processing from presentation frame processing. Please take a look.

**Comment by TrevorFSmith** [comment]
The pattern of the worker / processor seems like a good way to move CV out of the main thread.

For the iteration of cameras, I'm concerned that we're going to run into the same fingerprinting security concerns that are causing us to remove VRDevice iteration from the WebVR 2.0 spec. In the last implementers meeting the group decided to pursuit more of a constraints request pattern where the script asks for a device with some characteristics and the UA decides whether to respond with a device. Some browser venders are apparently unwilling to ship an API that allows the script iterate through all of the devices on a device, even after a prompt.

I believe that these security concerns are part of why the MediaDevices.getUserMedia call is constraint request based.

My other concern is that we're binding access to camera data to the XR APIs when they seem like a generic capability that may or may not be used in the context of XR. I can imagine many scenarios where a script would need the camera data but not want to display XR or sense the environment through SLAM. Should we consider extending the MediaDevices API and separately sync (or pre-process) camera data in conjunction with the XR frames?

**Comment by joshmarinacci** [comment]
Incidentally, I found that MediaDevices does have an iteration method as well, but I don't know if it's actually implemented.

https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/enumerateDevices <https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/enumerateDevices>

**Comment by huningxin** [comment]
> Incidentally, I found that MediaDevices does have an iteration method as well, but I don't know if it's actually implemented.

https://www.chromestatus.com/feature/4906859072847872

**Comment by huningxin** [comment]
> My other concern is that we're binding access to camera data to the XR APIs when they seem like a generic capability that may or may not be used in the context of XR. I can imagine many scenarios where a script would need the camera data but not want to display XR or sense the environment through SLAM. Should we consider extending the MediaDevices API and separately sync (or pre-process) camera data in conjunction with the XR frames?

It reminders me that we don't intent to solve a generic CV problem on Web. That might too big problem to solve. Let's limit the scope to: when the reality is a real world view (`reality.isPassthrough === true`) how developers access the captured image of the real world.

If that is the case, I guess developers can request the world images of a reality without enumerating the cameras. So I changed the `requestCameraFrame` to a method of `Reality`. Please take a look. @TrevorFSmith Thanks!

**Comment by blairmacintyre** [comment]
> For the iteration of cameras, I'm concerned that we're going to run into the same fingerprinting security concerns that are causing us to remove VRDevice iteration from the WebVR 2.0 spec. In the last implementers meeting the group decided to pursuit more of a constraints request pattern where the script asks for a device with some characteristics and the UA decides whether to respond with a device. Some browser venders are apparently unwilling to ship an API that allows scripts to iterate through all of the devices on a device, even after a prompt.
> 
> I believe that these security concerns are part of why the MediaDevices.getUserMedia call is constraint request based.

There is a big difference here:  cameras are integrated with devices, and on commercial HMDs will be fixed.  The Vive has it's camera;  the Hololens has it's cameras;  etc.  

So, after thinking about this, I believe that if the user/UA gives the program access to a given display, the programmer needs to see the camera structure (if they know what kind of display it is, they could already know what cameras are available, most likely).  As long as we only let the programmer iterate through the list of cameras for a display they have been given access to, we are in good shape.

**Comment by blairmacintyre** [comment]
> It reminders me that we don't intent to solve a generic CV problem on Web. That might too big problem to solve. Let's limit the scope to: when the reality is a real world view (reality.isPassthrough === true) how developers access the captured image of the real world.
> 
> If that is the case, I guess developers can request the world images of a reality without enumerating the cameras. So I changed the requestCameraFrame to a method of Reality. Please take a look. @TrevorFSmith Thanks!

I don't like this, and I don't think we should fall back to this approach.  A modern HMD may have many cameras attached and available to the programmer;  optical see-through displays may have cameras even if they aren't "pasthrough" (video mixed) devices.  We may want to do CV on cameras that aren't the "passthrough" camera on a display.

It's not even clear if future displays will "only" have a 1-1 mapping for video mixed displays.  Phones like the new iPhones already have multiple cameras on the back;  you can easily imagine passing through one to display and using both for CV.  Or, future phones with grids of cameras (like some of the lightfield cameras) may combine the video from the cameras in interesting ways, and present different interfaces to the programmer.

Please undo this change.

**Comment by blairmacintyre** [comment]
> My other concern is that we're binding access to camera data to the XR APIs when they seem like a generic capability that may or may not be used in the context of XR. I can imagine many scenarios where a script would need the camera data but not want to display XR or sense the environment through SLAM. Should we consider extending the MediaDevices API and separately sync (or pre-process) camera data in conjunction with the XR frames?

I think this is tricky.  On one hand I agree with you (I can see that use case, perhaps) but on the other hand, I'm not sure I see any compelling application of real-time CV on web-based platforms that wouldn't be satisfied by the ways of accessing cameras  that WebRTC provides.  WebRTC is  fine for non-real-time video access (where by real-time I mean the kinds of hard constraints AR/VR have, as opposed to looser application scenarios like video conferencing).

If we are talking about the range of sensors on AR/VR HMDs, accessing them within the context of one API makes sense:  the UA can manage access, and encapsulate a variety of data within permissions requests to the user (e.g., "give user access to display?  video?  depth? other sensors?" as part of one comprehensible permissions system, versus making the current mess of permissions even worse.

**Comment by TrevorFSmith** [comment]
Ok, I'm satisfied by the argument that if we're binding the security access of the display with access to the cameras on the display then iteration is fine. Also, that WebRTC is the appropriate place to get access to camera data for cameras not being used for XR.

Yes, we should plan for multiple cameras on displays because that is already the case for Microsoft MR HMDs and phones.

@huningxin Do you want to take another pass on the API or should I do it? It feels like we're mostly in agreement now.

**Comment by huningxin** [comment]
Thanks for the insights @blairmacintyre ! I revert the last change and fix some issues. 

@TrevorFSmith , please feel free to add more commits.

**Comment by TrevorFSmith** [comment]
I haven't forgotten about this PR (I've been traveling, sorry!) but in the meantime I noticed that the TAG review of the webvr spec mentioned Worklets from the CSS Painting API: https://www.w3.org/TR/css-paint-api-1/#paint-worklet

**Comment by huningxin** [comment]
That's interesting. Thanks for the notes!

Actually, I also noticed that web audio employs worklet for custom processing: https://webaudio.github.io/web-audio-api/#AudioWorklet. By using worklet, UA is free to determine running the worklet on main thread or worker thread. It also simplify the setup.

I pushed a commit to draft the API using worklet. @TrevorFSmith Please take a look.

**Comment by zoidhddhdd** [comment]
Zoid

**Pull request closed without merging**